### PR TITLE
Correct validateAuthInfo's logic of judgment

### PR DESCRIPTION
**Pull request by ping035627**
What type of PR is this?
/kind cleanup

What this PR does / why we need it:
The logic of validateAuthInfo looks a little confused. The verification of authInfo.ClientKey shouldn't be placed in the logic of authInfo.ClientCertificate. And it doesn’t verify that len(authInfo.ClientCertificate) == 0 && len(authInfo.ClientCertificateData) == 0 at the same time.

Which issue(s) this PR fixes:

Fixes #

Special notes for your reviewer:

Does this PR introduce a user-facing change?:

/release-note-none
/priority backlog

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/clientcmd/validation.go [hunk 1]
```diff
@@ -207,35 +207,39 @@ func validateAuthInfo(authInfoName string, authInfo clientcmdapi.AuthInfo) []err
 		methods = append(methods, "basicAuth")
 	}
 
-	if len(authInfo.ClientCertificate) != 0 || len(authInfo.ClientCertificateData) != 0 {
+	if len(authInfo.ClientCertificate) != 0 {
 		// Make sure cert data and file aren't both specified
```
Comment (xiangpengzhao) [inline staging/src/k8s.io/client-go/tools/clientcmd/validation.go]: Seems like the code comments aren't valid any more :)

**Review by xiangpengzhao — COMMENTED** [review]
Just some nits, otherwise LGTM. Ping @deads2k for another review :)

File: staging/src/k8s.io/client-go/tools/clientcmd/validation.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/clientcmd/validation.go]: please keep the same logic of ensuring that file and data can't both be specified

File: staging/src/k8s.io/client-go/tools/clientcmd/validation.go [hunk 2]
```diff
@@ -207,35 +207,39 @@ func validateAuthInfo(authInfoName string, authInfo clientcmdapi.AuthInfo) []err
...
-			validationErrors = append(validationErrors, fmt.Errorf("client-key-data or client-key must be specified for %v to use the clientCert authentication method.", authInfoName))
+
+		clientCertFile, err := os.Open(authInfo.ClientCertificate)
+		defer clientCertFile.Close()
+		if err != nil {
+			validationErrors = append(validationErrors, fmt.Errorf("unable to read client-cert %v for %v due to %v", authInfo.ClientCertificate, authInfoName, err))
 		}
+	}
+	// Make sure a key is specified
+	if len(authInfo.ClientCertificate) == 0 && len(authInfo.ClientCertificateData) == 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/clientcmd/validation.go]: this should not have been unnested... the previous logic checked that if client cert or certdata was specified, then either client key or keydata was also specified. this change makes it an error to not specify a client cert, which is incorrect.

**Comment by liggitt** [comment]
the validation was correct as is. client key or keydata is only required if client cert or cert data is specified

/close

**Pull request closed without merging**